btc/batcher.go

```diff
@@ -0,0 +1,524 @@
+	request, err := w.cache.ReadRequest(ctx, id)
+	if err != nil {
+		return Transaction{}, false, err
+	}
+	if !request.Status {
+		return Transaction{}, false, nil
+	}
+	batch, err := w.cache.ReadBatchByReqId(ctx, id)
+	if err != nil {
+		return Transaction{}, false, err
+	}
+
+	tx, err := w.indexer.GetTx(ctx, batch.Tx.TxID)
+	if err != nil {
+		return Transaction{}, false, err
+	}
+	return tx, true, nil
+}
+
+// Start starts the batcher wallet service
+func (w *batcherWallet) Start(ctx context.Context) error {
+	if w.quit != nil {
+		return ErrBatcherStillRunning
+	}
+	w.quit = make(chan struct{})
+
+	w.logger.Info("starting batcher wallet")
+	w.run(ctx)
+	return nil
+}
+
+// Stop gracefully stops the batcher wallet service
+func (w *batcherWallet) Stop() error {
+	if w.quit == nil {
+		return ErrBatcherNotRunning
+	}
+
+	w.logger.Info("stopping batcher wallet")
+	close(w.quit)
+	w.quit = nil
+
+	w.logger.Info("waiting for batcher wallet to stop")
+	w.wg.Wait()
+	return nil
+}
+
+// Restart restarts the batcher wallet service
+func (w *batcherWallet) Restart(ctx context.Context) error {
+	if err := w.Stop(); err != nil {
+		return err
+	}
+	return w.Start(ctx)
+}
+
+// starts the batcher based on the strategy
+// There are two types of batching triggers
+// 1. Periodic Time Interval (PTI) - Batches are created at regular intervals
+// 2. Pending Request - Batches are created when a certain number of requests are pending
+// 3. Exponential Time Interval (ETI) - Batches are created at exponential intervals but the interval is custom
+func (w *batcherWallet) run(ctx context.Context) {
+	switch w.opts.Strategy {
+	case CPFP, RBF:
+		w.runPTIBatcher(ctx)
+	default:
+		panic("strategy not implemented")
+	}
+}
+
+// PTI stands for Periodic time interval
+//  1. It creates a batch at regular intervals
+//  2. It also updates the fee rate at regular intervals
+//     if fee rate increases more than threshold and there are
+//     no batches to create
+func (w *batcherWallet) runPTIBatcher(ctx context.Context) {
+	w.wg.Add(1)
+	go func() {
+		defer w.wg.Done()
+		for {
+			ticker := time.NewTicker(w.opts.PTI)
+			select {
+			case <-w.quit:
+				return
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+				if err := w.createBatch(); err != nil {
+					if !errors.Is(err, ErrBatchParametersNotMet) {
+						w.logger.Error("failed to create batch", zap.Error(err))
+					} else {
+						w.logger.Info("waiting for new batch")
+					}
+
+					if err := w.updateFeeRate(); err != nil {
+						if !errors.Is(err, ErrFeeUpdateNotNeeded) {
+							w.logger.Error("failed to update fee rate", zap.Error(err))
+						} else {
+							w.logger.Info("fee update skipped")
+						}
+					} else {
+						w.logger.Info("batch fee updated", zap.String("strategy", string(w.opts.Strategy)))
+					}
+				} else {
+					w.logger.Info("new batch created", zap.String("strategy", string(w.opts.Strategy)))
+				}
```

Comment:
`createBatch` might fail for different reasons. Calling `updateFeeRate` on error might lead to corruption of data.